No hover is displayed if there is no documentation

datho7561 opened the pull request:
See https://github.com/redhat-developer/vscode-xml/issues/258

![no-doc-no-hover](https://user-images.githubusercontent.com/22376627/83181551-0cc5ad00-a0f3-11ea-87a2-c0bb0106a65b.gif)

Signed-off-by: David Thompson <davthomp@redhat.com>

angelozerr commented:
@datho7561 it should be better to fix this issue when documentation is generated. I have not tested but I think the correct fix is to move this block code https://github.com/eclipse/lemminx/blob/master/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java#L235-L248 to https://github.com/eclipse/lemminx/blob/03ca39622411febbf13b2721886985e0bcaef44c/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java#L233

For the test:

 * fix the tests by adding some xs:appinfo in docAppinfo.xsd
 * write test with empty documentation to check you return no hover.

datho7561 commented:
@angelozerr Thanks! The only issue is that this means the completion won't display anything if there is no doc:

![NoDocNoHoverLemminx](https://user-images.githubusercontent.com/22376627/83189137-bf4f3d00-a0fe-11ea-9d3d-86aeedd38115.png)

If this is okay, I'll take this route. Otherwise, I think something more involved might be required.

angelozerr commented:
Completion and Hover must have the same behavior to je consistent si yes its ok

fbricon commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java:
```diff
@@ -226,12 +227,14 @@ public static String generateAttributeValue(String defaultValue, Collection<Stri
 	 */
 	public static String generateDocumentation(String documentation, String schemaURI, boolean html) {
-		StringBuilder doc = new StringBuilder(documentation != null ? documentation : "");
+		if (StringUtil.isBlank(documentation)) {
```
Actually I think isBlank is more appropriate

angelozerr replied:
I would like to avoid having JSoup dependency in our code. Why isEmpty is not enough?

fbricon replied:
isBlank = isEmpty || only contains whitespaces. Use org.apache.commons.lang3.StringUtils instead

datho7561 replied:
Looking into how to properly filter out the whitespace. I think there is also html whitespace tags in documentation.

datho7561 replied:
I'm now using the lemminx `StringUtils.isEmpty`, but am also checking for all whitespace strings using `trim`.

fbricon commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java:
```diff
@@ -25,6 +25,7 @@
 import org.eclipse.lemminx.utils.XMLBuilder;
 import org.eclipse.lsp4j.MarkupContent;
 import org.eclipse.lsp4j.MarkupKind;
+import org.jsoup.helper.StringUtil;
```
use org.apache.commons.lang3.StringUtils that has isBlank

datho7561 commented:
> Given
> 
> ```
> <?xml version="1.0" encoding="UTF-8" ?>
> <hazdoc xmlns="foo" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="foo ./foo.xsd">
> <nodoc></nodoc>
> </hazdoc>
> ```
> 
> and foo.xsd containing
> 
> ```
> <?xml version="1.0"?>
> <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" 
>   elementFormDefault="qualified" 
>   xmlns="foo" 
>   targetNamespace="foo">
>   <xs:element name="hazdoc">
>     <xs:annotation>
>       <xs:appinfo> I haz appinfo</xs:appinfo>
>       <xs:documentation source="version">3.0.0+</xs:documentation>
>       <xs:documentation source="description">
>         
>         I haz doc
>         
>       </xs:documentation>
>     </xs:annotation>
>   </xs:element>
>       <xs:element name="nodoc" type="xs:string">
>         <xs:annotation>
>           <xs:appinfo>    </xs:appinfo>
>           <xs:documentation source="version">   </xs:documentation>
>           <xs:documentation source="description">
> 
>           </xs:documentation>
>         </xs:annotation>
>       </xs:element>
> </xs:schema>
> ```
> 
> The `nodoc` node should not have any hover, as all its docs are blank, yet it shows a Source link on hover, with no doc.

Thanks for point this out! I fixed it and added a test case for it.

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/utils/XMLGenerator.java:
```diff
@@ -226,12 +227,15 @@ public static String generateAttributeValue(String defaultValue, Collection<Stri
 	 */
 	public static String generateDocumentation(String documentation, String schemaURI, boolean html) {
-		StringBuilder doc = new StringBuilder(documentation != null ? documentation : "");
+		if (StringUtils.isEmpty(documentation) || StringUtils.isEmpty(documentation.trim())) {
```
Calling` documentation.trim()` will create a new instance of String which is shame just to check that documentation contains just whitespaces. 

Please try:

`if (StringUtils.isEmpty(documentation) || StringUtils.isWhitespace(documentation)) {`

datho7561 replied:
I added a isBlank function to StringUtils that checks `StringUtils.isEmpty(documentation) || StringUtils.isWhitespace(documentation)` and used this in both locations.

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/XSDDocumentation.java:
```diff
@@ -123,15 +123,17 @@ private static String getFormatted(String prefix, String elementName, List<Strin
-			} else {
-				result.append(System.lineSeparator());
-				result.append(System.lineSeparator());
+			if (!(StringUtils.isEmpty(doc) || StringUtils.isEmpty(doc.trim()))) {
```
`if (StringUtils.isEmpty(doc) || StringUtils.isWhitespace(doc)) {`

xorye commented on org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/XMLSchemaHoverDocumentationTypeTest.java:
```diff
@@ -301,6 +325,28 @@ private void assertElementMultipleBothHover(String expected, SchemaDocumentation
+	}
+
 	private void assertHover(String xml, String expected, SchemaDocumentationType docSource, boolean markdownSupported) throws BadLocationException {
 		String currSource = markdownSupported ? source : plainTextSource;
```
Could probably move this inside the if statement since the else block is not needed anymore

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/StringUtils.java:
```diff
@@ -63,6 +63,10 @@ public static boolean isWhitespace(String value) {
 		return isWhitespace(value, 0);
 	}
 
+	public static boolean isBlank(String value) {
```
Add Javadoc

fbricon commented:
In master, with:
```xml
<?xml version="1.0" standalone="no" ?>
<!DOCTYPE copyright [
  <!ELEMENT copyright (#PCDATA)>
  <!ENTITY c SYSTEM "http://www.xmlwriter.net/copyright.xml">
]>
<copyright>&c;</copyright>
```
Hovering over `&c;` I get:
<img width="569" alt="Screen Shot 2020-06-02 at 9 57 55 AM" src="https://user-images.githubusercontent.com/148698/83495134-9260a880-a4b7-11ea-9df9-97032de2dc7e.png">

WIth this PR, rebased against master, I get:

<img width="547" alt="Screen Shot 2020-06-02 at 9 52 42 AM" src="https://user-images.githubusercontent.com/148698/83495362-de135200-a4b7-11ea-91e8-915b16b69466.png">

angelozerr commented:
Works like a charm, great job @datho7561 !
